### PR TITLE
Parallel kmeans

### PR DESCRIPTION
**Pull request by BloodAxe**
This PR enables parallel processing of Kmean's internal routine, which should significantly improve 
kmeans performance on multi-core systems and big of trainset/number of clusters. 
A 1024 elements will be allocated per worker thread.

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
@BloodAxe unfortunately, your patch breakes http://pullrequest.opencv.org/buildbot/builders/precommit_opencl/builds/2362/steps/test_imgproc-ippicv-opencl/logs/tests%20summary

**Comment by BloodAxe** [comment]
Thanks for response. I'll investigate what went wrong with GC.

File: modules/core/src/kmeans.cpp [hunk 1]
```diff
@@ -212,6 +213,68 @@ class KMeansDistanceComputer : public ParallelLoopBody
...
+
+        const float * sample;
+        int j, k;
+
+        for (int i = begin; i < end; ++i)
+        {
+            sample = data.ptr<float>(i);
+            k = labels[i];
+
+            float* center = centers.ptr<float>(k);
```
Comment (ilya-lavrenov) [inline modules/core/src/kmeans.cpp]: if 2 threads process data with the same label, we get data races here

**Comment by vpisarev** [comment]
ok, let's close it for now; @BloodAxe, when you are ready to work on it further, please, reopen the PR

**Comment by BloodAxe** [comment]
@vpisarev sorry for abandoned PR. I was pretty busy, but I hope I will have some time in future to finish this feature.

**Pull request closed without merging**